### ISSUE TITLE
Branch Management for Features

### ISSUE THREAD

**Issue by zarfld** [zarfld/LinuxCnc_PokeysLibComp#70]
### Description:
To streamline development and ensure that features are properly tracked and managed, a well-defined branch management strategy should be implemented. Each new feature, bug fix, or improvement should have its own dedicated branch, which can then be reviewed and merged into the main branch. This helps avoid conflicts, allows for easier tracking of progress, and ensures that the main branch remains stable.

### Tasks:

1. **Define Branch Naming Conventions**:
   - Establish a clear naming convention for branches:
     - `feature/{feature-name}` for new features.
     - `bugfix/{bug-name}` for bug fixes.
     - `improvement/{improvement-name}` for general improvements or refactoring.
     - `hotfix/{hotfix-name}` for urgent fixes to be merged directly into the main branch.
   - Examples:
     - `feature/custom-homing-component`
     - `bugfix/digital-io-error`
     - `improvement/code-cleanup`

2. **Create Dedicated Branches for Each Feature**:
   - Each new feature or bug fix should have its own branch created from the latest version of the main branch.
   - The branch should focus on one feature or issue to ensure isolated development.

3. **Pull Request Process**:
   - Once development on a feature branch is complete, create a pull request (PR) to merge the changes into the main branch.
   - PRs should include:
     - A clear description of the changes made.
     - References to related issues (e.g., "Closes #123").
     - Automated tests, if applicable, to ensure the feature works as expected.
     - A checklist to confirm the feature is ready for review (e.g., "Code reviewed", "Tests passed", etc.).

4. **Review and Approval**:
   - Before a branch is merged into the main branch, it should go through a code review process.
   - Reviewers should check for:
     - Code quality and adherence to coding standards.
     - Potential conflicts or issues with existing code.
     - Proper documentation and unit tests.
   - Once approved, the branch can be merged.

5. **Rebasing and Conflict Resolution**:
   - Regularly rebase feature branches with the main branch to avoid conflicts during the merge process.
   - If conflicts arise, resolve them promptly and ensure that the feature branch is always up-to-date.

6. **Branch Cleanup**:
   - After a branch has been merged and the feature is live, delete the feature branch to keep the repository clean and organized.
   - Archive or document old branches if necessary to maintain a history of changes.

7. **Hotfix Branches**:
   - For urgent fixes that need to be applied directly to the main branch, create a hotfix branch.
   - After the hotfix is merged, ensure that any necessary changes are propagated to other branches as needed.

8. **Continuous Integration with Branches**:
   - Ensure that CI pipelines are set up to run automated tests for each branch to catch any potential issues early.
   - The CI pipeline should also handle deploying feature branches to a staging environment for further testing, if applicable.

### Acceptance Criteria:
- [ ] Branch naming conventions are defined and followed.
- [ ] Separate branches are created for each feature or bug fix.
- [ ] Pull requests are submitted and go through the review process before merging.
- [ ] Conflicts are resolved in a timely manner, and branches are kept up to date with the main branch.
- [ ] Feature branches are deleted after merging to keep the repository clean.
- [ ] CI pipelines are integrated with branches for automated testing and deployment.

### References:
- [GitHub Branch Management Best Practices](https://docs.github.com/en/get-started/quickstart/github-flow)
- [Gitflow Workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow)
